core/src/main/java/com/linecorp/armeria/server/DefaultRoute.java

```diff
@@ -312,6 +310,66 @@ public boolean isCacheable() {
         return paramPredicates.isEmpty() && headerPredicates.isEmpty() && excludedRoutes().isEmpty();
     }
 
+    @Override
+    public boolean hasDuplicateRouteCondition(Route other) {
+        if (complexity != other.complexity()) {
+            return false;
+        }
+
+        if (getClass() != other.getClass()) {
+            return false;
+        }
+
+        if (!pathType().hasTriePath() || !other.pathType().hasTriePath()) {
+            return false;
+        }
+
+        if (!paths().get(1).equals(other.paths().get(1))) {
+            // Just compare trie path.
+            return false;
+        }
+
+        if (methods.stream().noneMatch(method -> other.methods().contains(method))) {
+            // No overlap in supported methods.
+            return false;
+        }
+        if (!consumes.isEmpty() &&
+            consumes.stream().noneMatch(mediaType -> other.consumes().contains(mediaType))) {
+            // No overlap in consume types.
+            return false;
+        }
+        if (!produces.isEmpty() && produces.stream().noneMatch(
+                mediaType -> other.produces().contains(mediaType))) {
+            // No overlap in produce types.
+            return false;
+        }
+
```

Comment:
Question) I wonder if we should just return false if there are any custom type predicates - I think it's understandable that custom predicates cannot be compared anyways

e.g.
```java
if (paramPredicates.stream().anyMatch(RoutingPredicate::isCustomPredicate)) {
    return false;
}
```

Reply:
We consider two items to be duplicates if their consumer or producer types overlap in any way.
https://github.com/line/armeria/pull/5469/files#diff-77dc2cfe99f738b8b3fd5fa7197146fc98fd19436ed6bc27338cafa13470469aR174-R217

I used the same approach for the predicate as well.

Reply:
I see, I understood that we're basically looking for anything that matches - sounds good to me 👍